_source/_posts/2021-10-16-full-stack-java.md

```diff
@@ -0,0 +1,693 @@
+
+```shell
+export CYPRESS_E2E_USERNAME=<new-username>
+export CYPRESS_E2E_PASSWORD=<new-password>
+npm run e2e
+```
+
+Everything should pass in around a minute.
+
+```shell
+       Spec                                              Tests  Passing  Failing  Pending  Skipped
+  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
+  │ ✔  administration/administration.spec.      00:22        5        5        -        -        - │
+  │    ts                                                                                          │
+  └────────────────────────────────────────────────────────────────────────────────────────────────┘
+    ✔  All specs passed!                        00:22        5        5        -        -        -
+
+Execution time: 59 s.
+```
+
+## Create Entities to allow CRUD on Photos
+
+I've talked a lot about how to secure your application, but we haven't done anything with photos! JHipster has a JDL (JHipster Domain Language) feature that allows you to model the data in your app, and generate entities from it. You can use the [JDL Studio](https://start.jhipster.tech/jdl-studio/) to do this online and save it locally once you've finished.
+
+My data model for this app has `Album`, `Photo`, and `Tag` entities and sets up relationships between them. Below is a screenshot of what it looks like in JDL Studio.
+
+{% img blog/full-stack-java/jdl-studio.png alt:"JDL Studio" width:"800" %}{: .center-image }
+
+Copy the JDL below and save it in a `flickr2.jdl` file in the root directory of your project.
+
+```
+entity Album {
+  title String required
+  description TextBlob
+  created Instant
+}
+
+entity Photo {
+  title String required
+  description TextBlob
+  image ImageBlob required
+  height Integer
+  width Integer
+  taken Instant
+  uploaded Instant
+}
+
+entity Tag {
+  name String required minlength(2)
+}
+
+relationship ManyToOne {
+  Album{user(login)} to User
+  Photo{album(title)} to Album
+}
+
+relationship ManyToMany {
+  Photo{tag(name)} to Tag{photo}
+}
+
+paginate Album with pagination
+paginate Photo, Tag with infinite-scroll
+```
+
+You can generate entities and [CRUD](https://en.wikipedia.org/wiki/Create,_read,_update_and_delete) code (Java for Spring Boot; TypeScript and JSX for React) by using the following command:
+
+```shell
+jhipster jdl flickr2.jdl
+```
+
+When prompted, type `a` to allow overwriting of existing files.
+
+This process will create Liquibase changelog files (to create your database tables), entities, repositories, Spring MVC controllers, and all the React code that's necessary to create, read, update, and delete your entities. It'll even generate JUnit unit tests, Jest unit tests, and Cypress end-to-end tests!
+
+After the process completes, you can restart your app, log in, and browse through the **Entities** menu. Try adding some data to confirm everything works.
```

Comment:
well, I did what you said here, but unfortunately, I don't see any _Entities_ menu :( Where should it be visible?

Reply:
It should be in the top right, before Administration.

Reply:
Can you post the contents of your `flickr2.jdl` file and the output from running `jhipster jdl flickr2.jdl`. Maybe sure you're running this command from inside your app's directory.

FYI... the deployment to Heroku needs a fix for JHipster that's coming in v7.3.0. I plan to QA everything again once that's released. Hopefully it'll happen sometime this week.

Reply:
Sure, the content of flickr2.jdl is copypasted 1 to 1 from this article:
```
entity Album {
   title String required
   description TextBlob
   created Instant
}

entity Photo {
   title String required
   description TextBlob
   image ImageBlob required
   height Integer
   width Integer
   taken Instant
   uploaded Instant
}

entity Tag {
  name String required minlength(2)
}

relationship ManyToOne {
  Album{user(login)} to User
  Photo{album(title)} to Album
}

relationship ManyToMany {
  Photo{tag(name)} to Tag{photo}
}

paginate Album with pagination
paginate Photo, Tag with infinite-scroll

````

Here's the jhipster jdl output:
```
INFO! Using JHipster version installed locally in current project's node_modules
INFO! Executing import-jdl flickr2.jdl
INFO! The JDL is being parsed.
warn: An Entity name 'User' was used: 'User' is an entity created by default by JHipster. All relationships toward it will be kept but any attributes and relationships from it will be disregarded.
INFO! No change in entity configurations, no entities were updated.
INFO! The JDL has been successfully parsed
INFO! Generating 0 applications.
Congratulations, JHipster execution is complete!
Sponsored with ❤️  by @oktadev.

```
and here's a few last logs from starting the app:
```
2021-10-09 00:03:30.730 DEBUG 5912 --- [  restartedMain] com.auth0.flickr2.config.WebConfigurer   : Registering CORS filter
2021-10-09 00:03:30.768  INFO 5912 --- [  restartedMain] com.auth0.flickr2.config.WebConfigurer   : Web application configuration, using profiles: dev
2021-10-09 00:03:30.769 DEBUG 5912 --- [  restartedMain] com.auth0.flickr2.config.WebConfigurer   : Initialize H2 console
2021-10-09 00:03:30.778  INFO 5912 --- [  restartedMain] com.auth0.flickr2.config.WebConfigurer   : Web application fully configured
2021-10-09 00:03:31.042 DEBUG 5912 --- [  restartedMain] c.ehcache.core.Ehcache-usersByLogin      : Initialize successful.
2021-10-09 00:03:31.060 DEBUG 5912 --- [  restartedMain] c.ehcache.core.Ehcache-usersByEmail      : Initialize successful.
2021-10-09 00:03:31.063 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.User        : Initialize successful.
2021-10-09 00:03:31.065 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.Authority   : Initialize successful.
2021-10-09 00:03:31.067 DEBUG 5912 --- [  restartedMain] c.e.c.E.a.f.domain.User.authorities      : Initialize successful.
2021-10-09 00:03:31.069 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.Album       : Initialize successful.
2021-10-09 00:03:31.071 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.Photo       : Initialize successful.
2021-10-09 00:03:31.074 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.Photo.tags  : Initialize successful.
2021-10-09 00:03:31.076 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.Tag         : Initialize successful.
2021-10-09 00:03:31.078 DEBUG 5912 --- [  restartedMain] c.e.c.E.auth0.flickr2.domain.Tag.photos  : Initialize successful.
2021-10-09 00:03:31.090 DEBUG 5912 --- [  restartedMain] c.a.flickr2.config.AsyncConfiguration    : Creating Async Task Executor
2021-10-09 00:03:31.172 DEBUG 5912 --- [  restartedMain] c.a.f.config.LiquibaseConfiguration      : Configuring Liquibase
2021-10-09 00:03:31.382  WARN 5912 --- [flickr-2-task-1] t.j.c.liquibase.AsyncSpringLiquibase     : Starting Liquibase asynchronously, your database might not be ready at startup!
2021-10-09 00:03:32.150 DEBUG 5912 --- [flickr-2-task-1] t.j.c.liquibase.AsyncSpringLiquibase     : Liquibase has updated your database in 767 ms
2021-10-09 00:03:34.254 DEBUG 5912 --- [  restartedMain] t.j.c.apidoc.SpringfoxAutoConfiguration  : Starting OpenAPI docs
2021-10-09 00:03:34.265 DEBUG 5912 --- [  restartedMain] t.j.c.apidoc.SpringfoxAutoConfiguration  : Started OpenAPI docs in 11 ms
2021-10-09 00:03:34.296 DEBUG 5912 --- [  restartedMain] c.a.f.config.DatabaseConfiguration       : H2 database is available on port 18080
2021-10-09 00:03:34.536 DEBUG 5912 --- [  restartedMain] s.n.www.protocol.http.HttpURLConnection  : sun.net.www.MessageHeader@1335f5685 pairs: {GET /auth/realms/jhipster/.well-known/openid-configuration HTTP/1.1: null}{Accept: application/json, application/*+json}{User-Agent: Java/14}{Host: localhost:9080}{Connection: keep-alive}
2021-10-09 00:03:34.541 DEBUG 5912 --- [  restartedMain] s.n.www.protocol.http.HttpURLConnection  : sun.net.www.MessageHeader@4bdec35211 pairs: {null: HTTP/1.1 200 OK}{Cache-Control: no-cache, must-revalidate, no-transform, no-store}{X-XSS-Protection: 1; mode=block}{X-Frame-Options: SAMEORIGIN}{Referrer-Policy: no-referrer}{Date: Fri, 08 Oct 2021 22:03:34 GMT}{Connection: keep-alive}{Strict-Transport-Security: max-age=31536000; includeSubDomains}{X-Content-Type-Options: nosniff}{Content-Type: application/json}{Content-Length: 3829}
2021-10-09 00:03:34.555 DEBUG 5912 --- [  restartedMain] s.n.www.protocol.http.HttpURLConnection  : sun.net.www.MessageHeader@44fd25385 pairs: {GET /auth/realms/jhipster/protocol/openid-connect/certs HTTP/1.1: null}{User-Agent: Java/14}{Host: localhost:9080}{Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2}{Connection: keep-alive}
2021-10-09 00:03:34.562 DEBUG 5912 --- [  restartedMain] s.n.www.protocol.http.HttpURLConnection  : sun.net.www.MessageHeader@df9077211 pairs: {null: HTTP/1.1 200 OK}{Cache-Control: no-cache}{X-XSS-Protection: 1; mode=block}{X-Frame-Options: SAMEORIGIN}{Referrer-Policy: no-referrer}{Date: Fri, 08 Oct 2021 22:03:34 GMT}{Connection: keep-alive}{Strict-Transport-Security: max-age=31536000; includeSubDomains}{X-Content-Type-Options: nosniff}{Content-Type: application/json}{Content-Length: 1466}
2021-10-09 00:03:34.602 DEBUG 5912 --- [  restartedMain] jdk.event.security                       : X509Certificate: Alg:SHA256withRSA, Serial:17c61ebb785, Subject:CN=jhipster, Issuer:CN=jhipster, Key type:RSA, Length:2048, Cert Id:-1470580349, Valid from:08.10.2021, 23:58, Valid until:09.10.2031, 00:00
2021-10-09 00:03:35.202 DEBUG 5912 --- [  restartedMain] c.a.f.w.f.OAuth2RefreshTokensWebFilter   : Filter 'OAuth2RefreshTokensWebFilter' configured for use
2021-10-09 00:03:35.289  INFO 5912 --- [  restartedMain] org.jboss.threads                        : JBoss Threads version 3.1.0.Final
2021-10-09 00:03:35.686  INFO 5912 --- [  restartedMain] com.auth0.flickr2.Flickr2App             : Started Flickr2App in 8.686 seconds (JVM running for 9.2)
```
it seems like the entities are there but there are missing from the UI